flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointContextTest.java

```diff
@@ -0,0 +1,249 @@
+import org.apache.flink.runtime.concurrent.ManuallyTriggeredScheduledExecutor;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.executiongraph.Execution;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.ExecutionVertex;
+import org.apache.flink.runtime.executiongraph.failover.flip1.TestRestartBackoffTimeStrategy;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobgraph.JobVertex;
+import org.apache.flink.runtime.taskmanager.TaskExecutionState;
+import org.apache.flink.runtime.testtasks.NoOpInvokable;
+import org.apache.flink.runtime.testutils.CommonTestUtils;
+import org.apache.flink.runtime.testutils.DirectScheduledExecutorService;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.TestLogger;
+
+import org.apache.flink.shaded.guava18.com.google.common.collect.Sets;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * {@code StopWithSavepointContextTest} tests the stop-with-savepoint functionality of {@link
+ * SchedulerBase#stopWithSavepoint(String, boolean)}.
+ */
+public class StopWithSavepointContextTest extends TestLogger {
+
+    private JobGraph jobGraph;
+    private DefaultScheduler scheduler;
+
+    private StopWithSavepointOperations testInstance;
+
+    @Before
+    public void setup() throws Exception {
+        jobGraph = new JobGraph();
+
+        final JobVertex jobVertex = new JobVertex("vertex #0");
+        jobVertex.setInvokableClass(NoOpInvokable.class);
+        jobGraph.addVertex(jobVertex);
+
+        // checkpointInterval has to be set to a value lower than Long.MAX_VALUE to enable
+        // periodic checkpointing - only then can we enable/disable the CheckpointCoordinator
+        SchedulerTestingUtils.enablePeriodicCheckpointing(jobGraph, Long.MAX_VALUE - 1);
+        scheduler =
+                SchedulerTestingUtils.createSchedulerBuilder(
+                                jobGraph, ComponentMainThreadExecutorServiceAdapter.forMainThread())
+                        .setFutureExecutor(new DirectScheduledExecutorService())
+                        .build();
+        scheduler.startScheduling();
+
+        // the checkpoint scheduler is stopped before triggering the stop-with-savepoint
+        disableCheckpointScheduler();
+
+        testInstance = new StopWithSavepointContext(jobGraph.getJobID(), scheduler, this.log);
```

Comment:
This looks quite involved for a simple unit test. This could be an indicator that `StopWithSavepointContext` is not super easy to test yet.

Reply:
You're right. I refactored the `StopWithSavepointContextTest`. It uses `TestingSchedulerNG` now instead of `DefaultScheduler` which removes some complexity from the test code.